sql/moz-fx-data-shared-prod/udf_js/jackknife_percentile_ci/udf.sql

```diff
@@ -0,0 +1,247 @@
+/*
+
+Calculates a confidence interval using a jackknife resampling technique
+for a given percentile of a histogram. See
+https://en.wikipedia.org/wiki/Jackknife_resampling
+
+Users must specify the percentile of interest as the first parameter
+and a histogram struct as the second.
+
+*/
+CREATE OR REPLACE FUNCTION udf_js.jackknife_percentile_ci(
+  percentile FLOAT64,
+  histogram STRUCT<values ARRAY<STRUCT<key INT64, value FLOAT64>>>
+)
+RETURNS STRUCT<low FLOAT64, high FLOAT64, percentile FLOAT64> DETERMINISTIC
+LANGUAGE js
+AS
+  """
+  function computePercentile(percentile, histogram) {
+    if (percentile < 0 || percentile > 100) {
+      throw "percentile must be a value between 0 and 100";
+    }
+    let values = histogram.map(bucket => bucket.value);
+    let total = values.reduce((a, b) => a + b);
+    let normalized = values.map(value => value / total);
+    // Find the index into the cumulative distribution function that corresponds
+    // to the percentile. This undershoots the true value of the percentile.
+    let acc = 0;
+    let index = null;
+    for (let i = 0; i < normalized.length; i++) {
+      acc += normalized[i];
+      index = i;
+      if (acc >= percentile / 100) {
+        break;
+      }
+    }
+    // NOTE: we do not perform geometric or linear interpolation, but this would
+    // be the place to implement it.
+    return histogram[index].key;
+  }
+
+  function erfinv(x){
+      var z;
+      var a = 0.147;
+      var the_sign_of_x;
+      if(0==x) {
+          the_sign_of_x = 0;
+      } else if(x>0){
+          the_sign_of_x = 1;
+      } else {
+          the_sign_of_x = -1;
+      }
+      if(0 != x) {
+          var ln_1minus_x_sqrd = Math.log(1-x*x);
+          var ln_1minusxx_by_a = ln_1minus_x_sqrd / a;
+          var ln_1minusxx_by_2 = ln_1minus_x_sqrd / 2;
+          var ln_etc_by2_plus2 = ln_1minusxx_by_2 + (2/(Math.PI * a));
+          var first_sqrt = Math.sqrt((ln_etc_by2_plus2*ln_etc_by2_plus2)-ln_1minusxx_by_a);
+          var second_sqrt = Math.sqrt(first_sqrt - ln_etc_by2_plus2);
+          z = second_sqrt * the_sign_of_x;
+      } else { // x is zero
+          z = 0;
+      }
+      return z;
+  }
+
+  function array_sum(arr) {
+      return arr.reduce(function (acc, x) {
+          return acc + x;
+      }, 0);
+  }
+
+  function array_avg(arr) {
+      return array_sum(arr) / arr.length;
+  }
+
+  function getMeanErrorsPercentile(percentile, histogram, fullDataPercentile) {
+    var jk_percentiles = [];
+
+    histogram.values.forEach((bucket, i) => {
+      var histCopy = JSON.parse(JSON.stringify(histogram.values));
```

Comment:
Nit: I am not familiar with copying a histogram in JS. Is this converting to a string then parsing back optimal?

Reply:
This seems to be the most convenient way to create a deep-copy of the `histogram.values`. If we notice any impact on performance we could look into more efficient (which will also be a little more complex) alternatives.